docs/architecture.md

```diff
@@ -0,0 +1,139 @@
+
+By design, Go's [`crypto/x509`](https://pkg.go.dev/crypto/x509) library blocks
+non-standard certificate chains, such as [precertificate chains](https://www.rfc-editor.org/rfc/rfc6962#section-3.1)
+or chains that are not safe to use.
+TesseraCT uses [`internal/lax509`](/internal/lax509/), a lightweight fork of
+Go's [`crypto/x509`](https://pkg.go.dev/crypto/x509) library to parse chains.
+
+## Detailed architecture
+
+### Google Cloud Platform (GCP)
+
+This implementation is composed of:
+
+ 1. One or multiple TesseraCT servers. For reliability, multiple servers can run
+ concurently.
+ 1. Tesssera's backend infrastructure:
+    1. A [GCS](https://cloud.google.com/storage) bucket. TesseraCT re-uses this
+    bucket to store [issuer certificates](https://github.com/C2SP/C2SP/blob/main/static-ct-api.md#issuers).
+    1. A [Spanner](https://cloud.google.com/spanner) database used for
+    sequencing entries, and antispam.
+ 1. Key materials, stored in [Secret Manager](https://cloud.google.com/security/products/secret-manager)
+
+APIs:
+
+ 1. Submssion requests (`add-chain`, `add-pre-chain`, `get-roots`) are processed
+ by TesseraCT servers.
+ 2. Monitoring requests (fetching tiles, log entries and
+ issuers) are processed directly by GCS, without going through TesseraCT
+ servers.
+
+See Tessera [GCP design doc](https://github.com/transparency-dev/tessera/tree/main/storage/gcp)
+for additional details.
+
+### Amazon Web Services (AWS)
+
+This implementation is composed of:
+
+ 1. One or multiple TesseraCT servers. For reliability, multiple servers can run
+ concurently.
+ 1. Tesssera's backend infrastructure:
+    1. An [S3](https://aws.amazon.com/s3/) bucket. TesseraCT re-uses this bucket
+    to store [issuer certificates](https://github.com/C2SP/C2SP/blob/main/static-ct-api.md#issuers).
+    1. A MySQL [RDS](https://aws.amazon.com/rds/) database used for sequencing
+    entries, and antispam.
+ 1. Key materials, stored in [Secrets Manager](https://aws.amazon.com/secrets-manager/)
+
+APIs:
+
+ 1. Submssion requests (`add-chain`, `add-pre-chain`, `get-roots`) are processed
+by TesseraCT servers.
+ 2. Monitoring requests (fetching tiles, log entries and
+issuers) are processed directly by S3, without going through TesseraCT
+servers.
+
+See Tessera [GCP design doc](https://github.com/transparency-dev/tessera/tree/main/storage/gcp)
+for additional details.
+
+### S3-compliant systems with a MySQL database
+
+While TesseraCT's [AWS implementation](#amazon-web-services-aws)
+uses [RDS](https://aws.amazon.com/rds/) and [S3](https://aws.amazon.com/s3/),
+its configuration APIs _should_ be compatible with any MySQL database, and S3
+compatible backend, such as [MinIO](https://min.io/). This might be a good option
+to consider depending on your needs.
+
+> [!WARNING]
+> S3-compatible backends do not all provide the same guarantees
+> that S3 does, and might therefore not be suitable to run TesseraCT.
+
+### POSIX-compliant systems
+
```

Comment:
```suggestion

This implementation needs only:
1. A POSIX-compliant filesystem (e.g. ZFS) to store the log, and
1. Any HTTP server capable of directly serving the files from the log stored on that filesystem.

See the [POSIX readme](https://pkg.go.dev/github.com/transparency-dev/tessera/storage/posix) for more details.

If you are comfortable running a web-server and unix-style binaries in a VM (or on bare metal), and do not want the complexity of running additional storage and database services this might be a good option.
```